File: open_spiel/games/mfg/crowd_modelling.cc
// Copyright 2019 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/games/mfg/crowd_modelling.h"

#include <algorithm>
#include <cmath>
#include <memory>
#include <utility>
#include <vector>

#include "open_spiel/abseil-cpp/absl/strings/numbers.h"
#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
#include "open_spiel/abseil-cpp/absl/strings/str_join.h"
#include "open_spiel/abseil-cpp/absl/strings/str_split.h"
#include "open_spiel/abseil-cpp/absl/strings/substitute.h"
#include "open_spiel/spiel_utils.h"

namespace open_spiel {
namespace crowd_modelling {
namespace {
inline constexpr float kEpsilon = 1e-25;

// Facts about the game.
const GameType kGameType{/*short_name=*/"mfg_crowd_modelling",
                         /*long_name=*/"Mean Field Crowd Modelling",
                         GameType::Dynamics::kMeanField,
                         GameType::ChanceMode::kExplicitStochastic,
                         GameType::Information::kPerfectInformation,
                         GameType::Utility::kGeneralSum,
                         GameType::RewardModel::kRewards,
                         /*max_num_players=*/kNumPlayers,
                         /*min_num_players=*/kNumPlayers,
                         /*provides_information_state_string=*/true,
                         /*provides_information_state_tensor=*/false,
                         /*provides_observation_string=*/true,
                         /*provides_observation_tensor=*/true,
                         /*parameter_specification=*/
                         {{"size", GameParameter(kDefaultSize)},
                          {"horizon", GameParameter(kDefaultHorizon)}},
                         /*default_loadable*/true,
                         /*provides_factored_observation_string*/false};

std::shared_ptr<const Game> Factory(const GameParameters& params) {
  return std::shared_ptr<const Game>(new CrowdModellingGame(params));
}

std::string StateToString(int x, int t, Player player_id, bool is_chance_init) {
  if (is_chance_init) {
    return "initial";
  }
  if (player_id == 0) {
    return absl::Substitute("($0, $1)", x, t);
  }
  if (player_id == kMeanFieldPlayerId) {
    return absl::Substitute("($0, $1)_a", x, t);
  }
  if (player_id == kChancePlayerId) {
    return absl::Substitute("($0, $1)_a_mu", x, t);
  }
  SpielFatalError(absl::Substitute(
      "Unexpected state (player_id: $0, is_chance_init: $1)",
      player_id, is_chance_init));
}

REGISTER_SPIEL_GAME(kGameType, Factory);

}  // namespace

CrowdModellingState::CrowdModellingState(std::shared_ptr<const Game> game,
                                         int size, int horizon)
    : State(game),
      size_(size),
      horizon_(horizon),
      distribution_(size_, 1. / size_) {}

CrowdModellingState::CrowdModellingState(
    std::shared_ptr<const Game> game, int size, int horizon,
    Player current_player, bool is_chance_init, int x, int t, int last_action,
    double return_value, const std::vector<double>& distribution)
    : State(game),
      size_(size),
      horizon_(horizon),
      current_player_(current_player),
      is_chance_init_(is_chance_init),
      x_(x),
      t_(t),
      last_action_(last_action),
      return_value_(return_value),
      distribution_(distribution) {}

std::vector<Action> CrowdModellingState::LegalActions() const {
  if (IsTerminal()) return {};
  if (IsChanceNode()) return LegalChanceOutcomes();
  if (IsMeanFieldNode()) return {};
  SPIEL_CHECK_TRUE(IsPlayerNode());
  return {0, 1, 2};
}

ActionsAndProbs CrowdModellingState::ChanceOutcomes() const {
  if (is_chance_init_) {
    ActionsAndProbs outcomes;
    for (int i = 0; i < size_; ++i) {
      outcomes.push_back({i, 1. / size_});
    }
    return outcomes;
  }
  return {{0, 1. / 3}, {1, 1. / 3}, {2, 1. / 3}};
}

void CrowdModellingState::DoApplyAction(Action action) {
  SPIEL_CHECK_NE(current_player_, kMeanFieldPlayerId);
  return_value_ += Rewards()[0];
  if (is_chance_init_) {
    SPIEL_CHECK_GE(action, 0);
    SPIEL_CHECK_LT(action, size_);
    SPIEL_CHECK_EQ(current_player_, kChancePlayerId);
    x_ = action;
    is_chance_init_ = false;
    current_player_ = 0;
  } else if (current_player_ == kChancePlayerId) {
    x_ = (x_ + kActionToMove.at(action) + size_) % size_;
    ++t_;
    current_player_ = kMeanFieldPlayerId;
  } else {
    SPIEL_CHECK_EQ(current_player_, 0);
    x_ = (x_ + kActionToMove.at(action) + size_) % size_;
    last_action_ = action;
    current_player_ = kChancePlayerId;
  }
}

std::string CrowdModellingState::ActionToString(Player player,
                                                Action action) const {
  if (IsChanceNode() && is_chance_init_) {
    return absl::Substitute("init_state=$0", action);
  }
  return std::to_string(kActionToMove.at(action));
}

std::vector<std::string> CrowdModellingState::DistributionSupport() {
  std::vector<std::string> support;
  support.reserve(size_);
  for (int x = 0; x < size_; ++x) {
    support.push_back(StateToString(x, t_, kMeanFieldPlayerId, false));
  }
  return support;
}

void CrowdModellingState::UpdateDistribution(
    const std::vector<double>& distribution) {
  SPIEL_CHECK_EQ(current_player_, kMeanFieldPlayerId);
  SPIEL_CHECK_EQ(distribution.size(), size_);
  distribution_ = distribution;
  current_player_ = kDefaultPlayerId;
}

bool CrowdModellingState::IsTerminal() const { return t_ >= horizon_; }

std::vector<double> CrowdModellingState::Rewards() const {
  if (current_player_ != 0) {
    return {0.};
  }
  double r_x = 1 - 1.0 * std::abs(x_ - size_ / 2) / (size_ / 2);
  double r_a = -1.0 * std::abs(kActionToMove.at(last_action_)) / size_;
  double r_mu = -std::log(distribution_[x_]+kEpsilon);
  return {r_x + r_a + r_mu};
}

std::vector<double> CrowdModellingState::Returns() const {
  return {return_value_ + Rewards()[0]};
}

std::string CrowdModellingState::ToString() const {
  return StateToString(x_, t_, current_player_, is_chance_init_);
}

std::string CrowdModellingState::InformationStateString(Player player) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  return HistoryString();
}

std::string CrowdModellingState::ObservationString(Player player) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  return ToString();
}

void CrowdModellingState::ObservationTensor(Player player,
                                            absl::Span<float> values) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  SPIEL_CHECK_EQ(values.size(), size_ + horizon_ + 1);
  SPIEL_CHECK_LT(x_, size_);
  SPIEL_CHECK_GE(t_, 0);
  // Allow t_ == horizon_.
  SPIEL_CHECK_LE(t_, horizon_);
  std::fill(values.begin(), values.end(), 0.);
  if (x_ >= 0) {
    values[x_] = 1.;
  }
  // x_ equals -1 for the initial (blank) state, don't set any
  // position bit in that case.
  values[size_ + t_] = 1.;
}

std::unique_ptr<State> CrowdModellingState::Clone() const {
  return std::unique_ptr<State>(new CrowdModellingState(*this));
}

std::string CrowdModellingState::Serialize() const {
  std::string out =
      absl::StrCat(current_player_, ",", is_chance_init_, ",", x_, ",", t_, ",",
                   last_action_, ",", return_value_, "\n");
  absl::StrAppend(&out, absl::StrJoin(distribution_, ","));
  return out;
}

CrowdModellingGame::CrowdModellingGame(const GameParameters& params)
    : Game(kGameType, params),
      size_(ParameterValue<int>("size", kDefaultSize)),
      horizon_(ParameterValue<int>("horizon", kDefaultHorizon)) {}

std::vector<int> CrowdModellingGame::ObservationTensorShape() const {
  // +1 to allow for t_ == horizon.
  return {size_ + horizon_ + 1};
}

std::unique_ptr<State> CrowdModellingGame::DeserializeState(
    const std::string& str) const {
  std::vector<std::string> lines = absl::StrSplit(str, '\n');
  if (lines.size() != 2) {
    SpielFatalError(absl::StrCat("Expected 2 lines in serialized state, got: ",
                                 lines.size()));
  }
  Player current_player;
  int is_chance_init;
  int x;
  int t;
  int last_action;
  double return_value;
  std::vector<std::string> properties = absl::StrSplit(lines[0], ',');
  if (properties.size() != 6) {
    SpielFatalError(
        absl::StrCat("Expected 6 properties for serialized state, got: ",
                     properties.size()));
  }
  SPIEL_CHECK_TRUE(absl::SimpleAtoi(properties[0], &current_player));
  SPIEL_CHECK_TRUE(absl::SimpleAtoi(properties[1], &is_chance_init));
  SPIEL_CHECK_TRUE(absl::SimpleAtoi(properties[2], &x));
  SPIEL_CHECK_TRUE(absl::SimpleAtoi(properties[3], &t));
  SPIEL_CHECK_TRUE(absl::SimpleAtoi(properties[4], &last_action));
  SPIEL_CHECK_TRUE(absl::SimpleAtod(properties[5], &return_value));
  std::vector<std::string> serialized_distrib = absl::StrSplit(lines[1], ',');
  std::vector<double> distribution;
  distribution.reserve(serialized_distrib.size());
  for (std::string& v : serialized_distrib) {
    double parsed_weight;
    SPIEL_CHECK_TRUE(absl::SimpleAtod(v, &parsed_weight));
    distribution.push_back(parsed_weight);
  }
  return absl::make_unique<CrowdModellingState>(
      shared_from_this(), size_, horizon_, current_player, is_chance_init, x, t,
      last_action, return_value, distribution);
}

}  // namespace crowd_modelling
}  // namespace open_spiel

File: open_spiel/python/mfg/games/crowd_modelling.py
# Copyright 2019 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Lint as python3
"""Mean Field Crowd Modelling, implemented in Python.

This is a demonstration of implementing a mean field game in Python.

Fictitious play for mean field games: Continuous time analysis and applications,
Perrin & al. 2019 (https://arxiv.org/abs/2007.03458). This game corresponds
to the game in section 4.2.
"""

from typing import Any, List, Mapping
import numpy as np

from open_spiel.python.observation import IIGObserverForPublicInfoGame
import pyspiel

_NUM_PLAYERS = 1
_SIZE = 10
_HORIZON = 10
_NUM_ACTIONS = 3
_NUM_CHANCE = 3
_EPSILON = 10**(-25)
_DEFAULT_PARAMS = {"size": _SIZE, "horizon": _HORIZON}
_GAME_TYPE = pyspiel.GameType(
    short_name="python_mfg_crowd_modelling",
    long_name="Python Mean Field Crowd Modelling",
    dynamics=pyspiel.GameType.Dynamics.MEAN_FIELD,
    chance_mode=pyspiel.GameType.ChanceMode.EXPLICIT_STOCHASTIC,
    information=pyspiel.GameType.Information.PERFECT_INFORMATION,
    utility=pyspiel.GameType.Utility.GENERAL_SUM,
    reward_model=pyspiel.GameType.RewardModel.REWARDS,
    max_num_players=_NUM_PLAYERS,
    min_num_players=_NUM_PLAYERS,
    provides_information_state_string=True,
    provides_information_state_tensor=False,
    provides_observation_string=True,
    provides_observation_tensor=True,
    parameter_specification=_DEFAULT_PARAMS)


class MFGCrowdModellingGame(pyspiel.Game):
  """A Mean Field Crowd Modelling game.


    A game starts by an initial chance node that select the initial state
    of the MFG.
    Then the game sequentially alternates between:
      - An action selection node (Where the player Id >= 0)
      - A chance node (the player id is pyspiel.PlayerId.CHANCE)
      - A Mean Field node (the player id is pyspiel.PlayerId.MEAN_FIELD)
  """

  # pylint:disable=dangerous-default-value
  def __init__(self, params: Mapping[str, Any] = _DEFAULT_PARAMS):
    game_info = pyspiel.GameInfo(
        num_distinct_actions=_NUM_ACTIONS,
        max_chance_outcomes=max(params["size"], _NUM_CHANCE),
        num_players=_NUM_PLAYERS,
        min_utility=-np.inf,
        max_utility=+np.inf,
        utility_sum=0.0,
        max_game_length=params["horizon"])
    super().__init__(_GAME_TYPE, game_info, params)
    self.size = params["size"]
    self.horizon = params["horizon"]

  def new_initial_state(self):
    """Returns a state corresponding to the start of a game."""
    return MFGCrowdModellingState(self)

  def make_py_observer(self, iig_obs_type=None, params=None):
    """Returns an object used for observing game state."""
    if ((iig_obs_type is None) or
        (iig_obs_type.public_info and not iig_obs_type.perfect_recall)):
      return Observer(params, self)
    return IIGObserverForPublicInfoGame(iig_obs_type, params)

  def max_chance_nodes_in_history(self):
    """Maximun chance nodes in game history."""
    return self.horizon + 1


class MFGCrowdModellingState(pyspiel.State):
  """A Mean Field Crowd Modelling state."""

  # Maps legal actions to the corresponding move along the 1-D axis of the game.
  _ACTION_TO_MOVE = {0: -1, 1: 0, 2: 1}
  # Action that corresponds to no displacement.
  _NEUTRAL_ACTION = 1

  def __init__(self, game):
    """Constructor; should only be called by Game.new_initial_state."""
    super().__init__(game)
    self._is_chance_init = True  # is true for the first state of the game.
    self._player_id = pyspiel.PlayerId.CHANCE
    self._x = None
    self._t = 0
    # We initialize last_action to the neutral action. This makes sure
    # that the first reward does not include any displacement penalty.
    self._last_action = self._NEUTRAL_ACTION
    self.size = game.size
    self.horizon = game.horizon
    self.return_value = 0.0

    # Represents the current probability distribution over game states.
    # Initialized with a uniform distribution.
    self._distribution = [1. / self.size for i in range(self.size)]

  @property
  def x(self):
    return self._x

  @property
  def t(self):
    return self._t

  def state_to_str(self, x, t, player_id=pyspiel.PlayerId.DEFAULT_PLAYER_ID):
    """A string that uniquely identify a triplet x, t, player_id."""
    if self._is_chance_init:
      return "initial"
    if player_id == pyspiel.PlayerId.DEFAULT_PLAYER_ID:
      return str((x, t))
    if player_id == pyspiel.PlayerId.MEAN_FIELD:
      return str((x, t)) + "_a"
    if player_id == pyspiel.PlayerId.CHANCE:
      return str((x, t)) + "_a_mu"
    raise ValueError(
        "player_id is not mean field, chance or default player id.")

  # OpenSpiel (PySpiel) API functions are below. This is the standard set that
  # should be implemented by every perfect-information sequential-move game.

  def _legal_actions(self, player):
    """Returns a list of legal actions for player and MFG nodes."""
    if player == pyspiel.PlayerId.MEAN_FIELD:
      return []
    if (player == pyspiel.PlayerId.DEFAULT_PLAYER_ID
        and player == self.current_player()):
      return [0, 1, 2]
    raise ValueError(f"Unexpected player {player}. "
                     "Expected a mean field or current player 0.")

  def chance_outcomes(self):
    """Returns the possible chance outcomes and their probabilities."""
    if self._is_chance_init:
      return list(enumerate(self._distribution))
    return [(0, 1. / 3.), (1, 1. / 3.), (2, 1. / 3.)]

  def _apply_action(self, action):
    """Applies the specified action to the state."""
    if self._player_id == pyspiel.PlayerId.MEAN_FIELD:
      raise ValueError(
          "_apply_action should not be called at a MEAN_FIELD state.")
    self.return_value += self._rewards()
    if self._is_chance_init:
      # Here the action is between 0 and self.size - 1
      if action < 0 or action >= self.size:
        raise ValueError(
            "The action is between 0 and self.size - 1 at an init chance node")
      self._x = action
      self._is_chance_init = False
      self._player_id = 0
    elif self._player_id == pyspiel.PlayerId.CHANCE:
      # Here the action is between 0 and 2
      if action < 0 or action > 2:
        raise ValueError(
            "The action is between 0 and 2 at any chance node")
      self._x = (self.x + self._ACTION_TO_MOVE[action]) % self.size
      self._t += 1
      self._player_id = pyspiel.PlayerId.MEAN_FIELD
    elif self._player_id == 0:
      # Here the action is between 0 and 2
      if action < 0 or action > 2:
        raise ValueError(
            "The action is between 0 and 2 at any chance node")
      self._x = (self.x + self._ACTION_TO_MOVE[action]) % self.size
      self._last_action = action
      self._player_id = pyspiel.PlayerId.CHANCE

  def _action_to_string(self, player, action):
    """Action -> string."""
    del player
    if self.is_chance_node() and self._is_chance_init:
      return f"init_state={action}"
    return str(self._ACTION_TO_MOVE[action])

  def distribution_support(self):
    """return a list of state string."""
    return [
        self.state_to_str(
            i, self.t, player_id=pyspiel.PlayerId.MEAN_FIELD)
        for i in range(self.size)
    ]

  def update_distribution(self, distribution):
    """This function is central and specific to the logic of the MFG.

    Args:
      distribution: a distribution to register.

      - function should be called when the node is in MEAN_FIELD state.
      - distribution are probabilities that correspond to each game state
      given by distribution_support.

    """
    if self._player_id != pyspiel.PlayerId.MEAN_FIELD:
      raise ValueError(
          "update_distribution should only be called at a MEAN_FIELD state.")
    self._distribution = distribution.copy()
    self._player_id = pyspiel.PlayerId.DEFAULT_PLAYER_ID

  def is_terminal(self):
    """Returns True if the game is over."""
    return self.t >= self.horizon

  def current_player(self):
    """Returns id of the next player to move, or TERMINAL if game is over."""
    if self.is_terminal():
      return pyspiel.PlayerId.TERMINAL
    return self._player_id

  def _rewards(self):
    """Reward for the player for this state."""
    if self._player_id == 0:
      r_x = 1 - (1.0 * np.abs(self.x - self.size // 2)) / (self.size // 2)
      r_a = -(1.0 * np.abs(self._ACTION_TO_MOVE[self._last_action])) / self.size
      r_mu = - np.log(self._distribution[self.x] + _EPSILON)
      return r_x + r_a + r_mu
    return 0.0

  def rewards(self) -> List[float]:
    """Rewards for all players."""
    # For now, only single-population (single-player) mean field games
    # are supported.
    return [self._rewards()]

  def _returns(self):
    """Returns is the sum of all payoffs collected so far."""
    return self.return_value + self._rewards()

  def returns(self) -> List[float]:
    """Returns for all players."""
    # For now, only single-population (single-player) mean field games
    # are supported.
    return [self._returns()]

  def __str__(self):
    """A string that uniquely identify the current state."""
    return self.state_to_str(self.x, self.t, player_id=self._player_id)


class Observer:
  """Observer, conforming to the PyObserver interface (see observation.py)."""

  def __init__(self, params, game):
    """Initializes an empty observation tensor."""
    del params

    self.size = game.size
    self.horizon = game.horizon
    # +1 to allow t == horizon.
    self.tensor = np.zeros(self.size + self.horizon + 1, np.float32)
    self.dict = {"x": self.tensor[:self.size], "t": self.tensor[self.size:]}

  def set_from(self, state: MFGCrowdModellingState, player: int):
    """Updates `tensor` and `dict` to reflect `state` from PoV of `player`."""
    del player
    # We update the observation via the shaped tensor since indexing is more
    # convenient than with the 1-D tensor. Both are views onto the same memory.
    self.tensor.fill(0)
    # state.x is None for the initial (blank) state, don't set any
    # position bit in that case.
    if state.x is not None:
      if not 0 <= state.x < self.size:
        raise ValueError(
            f"Expected {state} x position to be in [0, {self.size})")
      self.dict["x"][state.x] = 1
    if not 0 <= state.t <= self.horizon:
      raise ValueError(f"Expected {state} time to be in [0, {self.horizon}]")
    self.dict["t"][state.t] = 1

  def string_from(self, state, player):
    """Observation of `state` from the PoV of `player`, as a string."""
    del player
    return str(state)


# Register the game with the OpenSpiel library

pyspiel.register_game(_GAME_TYPE, MFGCrowdModellingGame)
